src/ucs/datastruct/conn_match.h

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_CONN_MATCH_H_
+#define UCS_CONN_MATCH_H_
+
+#include <ucs/datastruct/khash.h>
+#include <ucs/datastruct/hlist.h>
+
+#include <inttypes.h>
+
+
+/**
+ * Maximal length of address
+ */
+#define UCS_CONN_MATCH_ADDR_STR_MAX    128
+
+
+/**
+ * Connection sequence number
+ */
+typedef uint16_t ucs_conn_sn_t;
+
+
+/**
+ * Connection address
+ */
+typedef struct ucs_conn_match_addr {
+    void                     *addr;           /* Pointer to the destination address */
+    size_t                   length;          /* Length of the destination address */
+} ucs_conn_match_addr_t;
+
+
+/**
+ * Connection queue type
+ */
+typedef enum ucs_conn_match_queue_type {
+    /* Queue type for connections created by API and not
+     * connected to remote peer */
+    UCS_CONN_MATCH_QUEUE_EXP,
+    /* Queue type for connections created internally as
+     * connected to remote peer, but not provided to user yet */
+    UCS_CONN_MATCH_QUEUE_UNEXP
+} ucs_conn_match_queue_type_t;
+
+
+/**
+ * Structure to embed in a connection entry to support matching with remote
+ * peer's connections
+ */
+typedef struct ucs_conn_match {
+    ucs_hlist_link_t          list;          /* List entry into endpoint
+                                                matching structure */
+} ucs_conn_match_t;
+
+
+/**
+ * Connection entry - allows *ordered* matching of connections between a pair
+ * of connected peers. The expected/unexpected lists are *not* circular
+ */
+typedef struct ucs_conn {
+    ucs_hlist_head_t         conn_q[2];       /* Connection queues */
+    ucs_conn_sn_t            next_conn_sn;    /* Sequence number of matching
+                                                 connections, since wireup messages
+                                                 used for connection establishment
+                                                 procedure which were sent on different
+                                                 connections could be provided
+                                                 out-of-order */
+} ucs_conn_t;
```

Comment:
ucs_conn_t -> ucs_conn_match_peer_t (all object in this header must begin with ucs_conn_match_ prefix)